packages/session-replay-browser/src/helpers.ts

```diff
@@ -1,13 +1,102 @@
 import { getGlobalScope } from '@amplitude/analytics-client-common';
-import { UNMASK_TEXT_CLASS } from './constants';
+import { MASK_TEXT_CLASS, UNMASK_TEXT_CLASS } from './constants';
+import { DEFAULT_MASK_LEVEL, MaskLevel, PrivacyConfig } from './config/types';
+import { getInputType } from '@amplitude/rrweb-snapshot';
 
-export const maskInputFn = (text: string, element: HTMLElement) => {
-  if (element.classList?.contains(UNMASK_TEXT_CLASS)) {
-    return text;
+// compiler guard
+/* istanbul ignore next */
+const assertUnreachable = (_: never): never => {
+  throw new Error('did not expect to get here');
+};
+
+/**
+ * Light: Subset of inputs
+ * Medium: All inputs
+ * Conservative: All inputs and all texts
+ */
+const isMaskedForLevel = (element: HTMLElement, elementType: 'input' | 'text', level: MaskLevel): boolean => {
+  switch (level) {
+    case MaskLevel.LIGHT: {
+      if (elementType !== 'input') {
+        return false;
+      }
+
+      const inputType = getInputType(element);
+      /* istanbul ignore if */ // For some reason it's impossible to test this.
+      if (!inputType) {
+        return false;
+      }
+
+      if (['password', 'hidden', 'email', 'tel'].includes(inputType)) {
+        return true;
+      }
+
+      if ((element as HTMLInputElement).autocomplete.startsWith('cc-')) {
+        return true;
+      }
+
+      return false;
+    }
+    case MaskLevel.MEDIUM:
+      return elementType === 'input';
+    case MaskLevel.CONSERVATIVE:
+      return elementType === 'input' || elementType === 'text';
   }
-  return '*'.repeat(text.length);
+  /* istanbul ignore next */
+  return assertUnreachable(level);
 };
 
+/**
+ * Checks if the given element set to be masked by rrweb
+ *
+ * Priority is:
+ *  1. [In code] Element/class based masking/unmasking
+ *  2. [Config based] Selector based masking/unmasking
+ *  3. Use app defaults
+ */
+const isMasked = (element: HTMLElement, elementType: 'input' | 'text', config?: PrivacyConfig): boolean => {
+  // Element is explicitly instrumented to mask/unmask
+  if (element.classList) {
+    if (element.classList.contains(MASK_TEXT_CLASS)) {
```

Comment:
i think the initial proposal was the most restrictive would win (out of 1&2).

Reply:
That's right. Just to confirm though do you mean 1&2 are similar priority, but prefer masking or unmasking that is?

Reply:
Yea, mostly thinking about the case where the engineer and admin configures something conflicting. We err on masking and let them figure it out.

Reply:
Yep, makes sense!